src/java.naming/share/classes/com/sun/jndi/ldap/pool/Pool.java

```diff
@@ -117,45 +120,109 @@ public Pool(int initSize, int prefSize, int maxSize) {
     public PooledConnection getPooledConnection(Object id, long timeout,
         PooledConnectionFactory factory) throws NamingException {
 
+        final long start = System.nanoTime();
+        long remaining = timeout;
+
         d("get(): ", id);
         if (debug) {
             synchronized (map) {
                 d("size: ", map.size());
             }
+            remaining = checkRemaining(start, remaining);
         }
 
         expungeStaleConnections();
 
+        Connections conns = getOrCreateConnections(factory, id);
+        d("get(): size after: ", map.size());
+        remaining = checkRemaining(start, remaining);
+
+        if (!conns.grabLock(remaining)) {
+            throw new NamingException("Timed out waiting for lock");
```

Comment:
Is this the appropriate exception? I see in `checkRemaining`:
```
                throw new CommunicationException(
                        "Timeout exceeded while waiting for a connection: " +
                                timeout + "ms");
```

Reply:
I've changed this to a CommuncationException.